stdlib/asm/crypto/stark/deep_queries.masm

```diff
@@ -3,782 +3,388 @@ const.MAIN_TRACE_COM_PTR=4294967290
+    #=> [(i-w)1, (i-w)0, (j-q)1, (j-q)0, CURRENT_TRACE_ROW_PTR, OOD_TRACE_PTR + 1, DEEP_RAND_CC_PTR, 0, ...]
 
     # Multiply by randomness
-    ## Get [alpha1, alpha0, beta1, beta0]
-    dup.6
-    add.1
-    swap.7
-    mem_loadw
-    #=> [[alpha1, alpha0, beta1, beta0], value_i, p1, p2, p3, T, Acc]
-    swapdw
-    movupw.2
-    #=> [[alpha1, alpha0, beta1, beta0], T, Acc, value_i, p1, p2, p3]
-
-    ## Multiply shifted term
-    dup.7
-    dup.7
+
+    ## Multiply (i - w) first and move it to its expected place on the stack
+    push.0.0.0.0
+    dup.10 mem_loadw drop drop 
+    #=> [u1, u0, (i-w)1, (i-w)0, (j-q)1, (j-q)0, CURRENT_TRACE_ROW_PTR, OOD_TRACE_PTR + 1, DEEP_RAND_CC_PTR, 0, ...]
     ext2mul
-    #=> [prod1_g, prod0_g, beta1, beta0, - T_i(z)_1, value_i - T_i(z)_0, - T_i(gz)_1, value_i - T_i(gz)_Acc, 0, value_i, p1, p2, p3]
-    # Where (prod0_g, prod1_g) =  (value_i - T_i(gz)_0, - T_i(gz)_1) * alpha = (T_i(x) - T_i(gz)) * alpha
-    swap
-    swap.3
-    swap
-    swap.2
-    swapw
-    #=> [- T_i(z)_1, value_i - T_i(z)_0, - T_i(gz)_1, value_i - T_i(gz)_0, beta1, beta0, prod1_g, prod0_g, Acc, value_i, p1, p2, p3]
+    movdn.3 movdn.3
+    #=> [(j-q)1, (j-q)0, i1', i0', CURRENT_TRACE_ROW_PTR, OOD_TRACE_PTR + 1, DEEP_RAND_CC_PTR, 0, ...]
 
-    ## Multiply non-shifted term
-    dup.5
-    dup.5
+    ## Multiply (j - q)
+    push.0.0.0.0
+    dup.10 add.2 swap.11 add.1 mem_loadw drop drop
+    #=> [v1, v0, (j-q)1, (j-q)0, i1', i0', CURRENT_TRACE_ROW_PTR, OOD_TRACE_PTR + 1, DEEP_RAND_CC_PTR + 2, 0, ...]
     ext2mul
-    #=> [prod1, prod0, - T_i(gz)_1, value_i - T_i(gz)_0, beta1, beta0, prod1_g, prod0_g, Acc, value_i, p1, p2, p3]
-    # where (prod0, prod1) =  (value_i - T_i(z)_0, - T_i(z)_1) * beta = (T_i(x) - T_i(z)) * beta
-    movup.7
-    movup.7
-    dupw.2
-    #=> [Acc, prod1_g, prod0_g, prod1, prod0, y, y, y, y, Acc, value_i, p1, p2, p3]
-
-    # Accumulate
-    movup.7
-    movup.7
-    ext2add
-    movdn.5
-    movdn.5
-    ext2add
-    swap
-    swap.3
-    swap
-    swap.2
-    #=> [Acc3 + prod1, Acc2 + prod0, Acc1 + prod1_g, Acc0 + prod0_g, y, y, y, y, Acc, value_i, p1, p2, p3]           Acc = [Tz, Tz, Tgz, Tgz]
-    #=> [Acc, Y, Y, P]
-
-    # Prepare for next cycle, if any
-    swapw.3
-    swapw
-    #=> [Y, P, Y, Acc]
+    #=> [j1', j0', i1', i0', CURRENT_TRACE_ROW_PTR, OOD_TRACE_PTR + 1, DEEP_RAND_CC_PTR + 2, 0, ...]
 end
 
-
-
-#! Compute the deep composition polynomial at the previously drawn random queries. Currently, their
-#! number is fixed to 27.
+#! Compute the deep composition polynomial at the previously drawn random queries.
 #!
-#! Input: [...]
+#! Input: [num_queries, ...]
 #! Output: [...]
-#! Cycles: 44 + 27 * 8445 = 228059
+#! Cycles: 17 + num_queries * 590
 export.compute_deep_composition_polynomial_queries
     push.FRI_QUERIES_PTR
-    repeat.5 push.0.0.0.0 end
+    add
     push.FRI_QUERIES_PTR
-    #=>[ptr_query, Y, Y, Y, Y, Y, ptr_query]
+    swap
+    push.0.0.0.0
+    push.FRI_QUERIES_PTR
```

Comment:
minor nit: I'd probably do it like this
```
push.FRI_QUERIES_PTR
swap
dup.1
add
padw
dup.5
```
It is the same number of cycles but uses fewer `push` operations.

Reply:
Thanks!